lib/Sema/TypeCheckEffects.cpp

```diff
@@ -3569,11 +3569,26 @@ class CheckEffectsCoverage : public EffectsHandlingWalker<CheckEffectsCoverage>
       return;
     const std::vector<DiagnosticInfo> &errors = asyncPointIter->getSecond();
     SourceLoc awaitInsertLoc = anchor->getStartLoc();
+    std::string fixIt = "await ";
     if (const AnyTryExpr *tryExpr = dyn_cast<AnyTryExpr>(anchor))
       awaitInsertLoc = tryExpr->getSubExpr()->getStartLoc();
     else if (const AutoClosureExpr *autoClosure = dyn_cast<AutoClosureExpr>(anchor)) {
       if (const AnyTryExpr *tryExpr = dyn_cast<AnyTryExpr>(autoClosure->getSingleExpressionBody()))
         awaitInsertLoc = tryExpr->getSubExpr()->getStartLoc();
+    } else {
+      // Check if we are looking at a shorthand optional binding.
+      // In that case, we'll want to supply a different fixIt.
+      const DeclRefExpr *declRefExpr;
+      if (const LoadExpr *loadExpr = dyn_cast<LoadExpr>(anchor)) {
+        declRefExpr = dyn_cast<DeclRefExpr>(loadExpr->getSubExpr());
+      } else {
+        declRefExpr = dyn_cast<DeclRefExpr>(anchor);
+      }
+      if (declRefExpr && declRefExpr->isImplicit()) {
+        if (const VarDecl *varDecl = dyn_cast<VarDecl>(declRefExpr->getDecl())) {
+          fixIt = (varDecl->getNameStr() + " = await ").str();
+        }
+      }
```

Comment:
`LoadExpr` is not the only thing that can get in the way:

```swift
class A {}
class B: A {}

func test() async {
  async let result: B? = nil
  if let result: A {}
}
```

Implicitness is a necessary condition for shorthand optional binding here, but I am not quite as certain it is sufficient too. You may have noticed by now that inserting `result = await` before a legal typed pattern as above will not work. The general solution is to insert *after* the pattern, so it seems we need to propagate the pattern one way or another for both a definitive check and the correct insertion location.

Reply:
@AnthonyLatsis I was able to find a better solution, you'd given me a good hint. Even still, it took me a while to realize we probably have other `fixItInsert` functions (like ones that can insert _after_). The new solution is to use `fixItInsertAfter` with the specialized fixIt given the conditions:
1) The expression type is optional.
2) Has a sub expression that is a DeclRefExpr and Implicit.


My reasoning is that I saw in ParseStmt.cpp, where we parse a shorthand optional binding we have:
```swift
else if (!ThePattern.getPtrOrNull()->getBoundName().empty()) {
    auto bindingName = DeclNameRef(ThePattern.getPtrOrNull()->getBoundName());
    auto loc = DeclNameLoc(ThePattern.getPtrOrNull()->getEndLoc());
    auto declRefExpr = new (Context) UnresolvedDeclRefExpr(bindingName,
                                                           DeclRefKind::Ordinary,
                                                           loc);
    
     declRefExpr->setImplicit();
     Init = makeParserResult(declRefExpr);
```
And I am thinking that implies 1 and 2, though I'm still unsure if it holds in the other direction.

Reply:
Hi @AnthonyLatsis, could you take another look at this for me when you get the chance?

Reply:
Hello, James. I’m sorry it took me so long to get back to you. What happened to the changes?

Reply:
Hi, sorry I was updating my checkout last night and switched the branch that I have this PR on without realizing it would close this. I reopened it with the changes on https://github.com/apple/swift/pull/72726.